### PR TITLE
Support systems with both glibc and Musl

### PR DESCRIPTION
**Pull request by graebm**
**Issue #:** https://github.com/awslabs/aws-crt-java/issues/659

Customer has a system with both glibc and Musl on it. The system is Alpine-based, and uses Musl for most things. But the installed Java is using glibc.

aws-crt-java used to work on this system, back when the shared lib only worked with glibc. But as of [v0.23.0](https://github.com/awslabs/aws-crt-java/releases/tag/v0.23.0) we have different shared libs for glibc and Musl, and need to detect at runtime which to use.

Our current method checks only checks whether the system's built-in libraries use Musl.

**Description of changes:**

NEW: Let users specify "musl" or "glibc" via system property "aws.crt.libc".

NEW: Otherwise, check whether the current Java executable is using glibc or Musl.

And if those fail, fall back to the previous method of calling `ldd --version` to see if it mentions "musl".

**Testing:**

I tested this by hand. It would be a ton of effort to hook up regular CI for this. Seeing as this is such an edge-case, that feels sufficient.

I didn't touch the compile-time CMake logic, only the runtime logic. So you can't test this without using our released JARs (which have shared libs for every platform), or by preparing a JAR by hand with shared libs for both Musl and glibc. Again, since this is such an edge case, this level of effort seems sufficient.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by waahm7 — APPROVED** [review]
fix & ship

File: src/main/java/software/amazon/awssdk/crt/CRT.java [hunk 1]
```diff
@@ -152,40 +153,83 @@ public static String getCRuntime(String osIdentifier) {
...
-                outputBuilder.append(line);
+        // The system might have both musl and glibc on it:
+        // https://github.com/awslabs/aws-crt-java/issues/659
+
+        // First, check which one java is using.
+        // Run: ldd $JAVA_HOME/bin/java
+        // If musl, has a line like: libc.musl-x86_64.so.1 => /lib/ld-musl-x86_64.so.1 (0x7f7732ae4000)
+        // If glibc, has a line like: libc.so.6 => /lib64/ld-linux-x86-64.so.2 (0x7f112c894000)
+        Pattern muslWord = Pattern.compile("\\bmusl\\b", Pattern.CASE_INSENSITIVE);
+        Pattern libcWord = Pattern.compile("\\blibc\\b", Pattern.CASE_INSENSITIVE);
```
[the remark this answers is not in the document]
Comment (graebm) [reply src/main/java/software/amazon/awssdk/crt/CRT.java:164]: There are N+1 Linux distributions out there.
I guarantee someone's working on new distro now just because they wish "musl" was capitalized differently.

File: src/main/java/software/amazon/awssdk/crt/CRT.java [hunk 2]
```diff
@@ -152,40 +153,83 @@ public static String getCRuntime(String osIdentifier) {
...
-
-            // sometimes, ldd's output goes to stderr, so capture that too
-            BufferedReader stdError = new BufferedReader(new
-                    InputStreamReader(proc.getErrorStream()));
-
-            String line;
-            StringBuilder outputBuilder = new StringBuilder();
-            while ((line = stdOutput.readLine()) != null) {
-                outputBuilder.append(line);
+        // The system might have both musl and glibc on it:
```
[the remark this answers is not in the document]
Comment (graebm) [reply src/main/java/software/amazon/awssdk/crt/CRT.java:156]: Currently, CRT java code logs via our native logger, but we can't use that yet because the C code isn't loaded...

The only place we do `System.err.println()` is when the CRT has failed to load.

I'm hesitant to print debug-but-not-error info to stderr or stdout. People hate when libraries do that.

I'm just going to 🤷‍♀️ for now

File: src/main/java/software/amazon/awssdk/crt/CRT.java [hunk 3]
```diff
@@ -152,40 +153,83 @@ public static String getCRuntime(String osIdentifier) {
...
+        // https://github.com/awslabs/aws-crt-java/issues/659
+
+        // First, check which one java is using.
+        // Run: ldd $JAVA_HOME/bin/java
```
[the remark this answers is not in the document]
Comment (graebm) [reply src/main/java/software/amazon/awssdk/crt/CRT.java:160]: The comment was a hand-wavey allusion to something you could copy paste to your own terminal
I changed it to be less literal by pretty obvious `/path/to/java`

[the remark this answers is not in the document]
Comment (graebm) [reply src/main/java/software/amazon/awssdk/crt/CRT.java:160]: Yeah, that's a solid idea. Now user's can do `-Daws.crt.libc=musl` or `-Daws.crt.libc=glibc` if our autodetection fails.

I ended up documenting all our System properties in the README, since none of them were documented, and I needed to document the new one somewhere

[the remark this answers is not in the document]
Comment (graebm) [reply src/main/java/software/amazon/awssdk/crt/CRT.java]: And yeah, there are lots of other ways to detect Musl out there, and every PR and stackoverflow  has people commenting "this still doesn't work on my machine". We can revisit other ways in the future. Just run them all one after another. But this seems good enough for now...
